builtin/gc.c

```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
 	return 0;
 }
 
+static void link_or_copy_or_die(const char *src, const char *dst)
+{
+	if (!link(src, dst))
+		return;
+
+	warning_errno(_("failed to link '%s' to '%s'"), src, dst);
+
+	if (!copy_file(dst, src, 0644))
+		return;
+
+	die_errno(_("failed to copy '%s' to '%s'"), src, dst);
+}
+
+static void migrate_pack(const char *srcdir, const char *dstdir,
+			 const char *pack_filename)
+{
+	struct stat st;
+	char *basename;
+	struct strbuf src = STRBUF_INIT, dst = STRBUF_INIT;
+	struct {
+		const char *ext;
+		unsigned move:1;
+	} files[] = {
+		{".pack", 0},
+		{".keep", 0},
+		{".rev", 0},
+		{".idx", 1}, /* The index file must be atomically moved last. */
+	};
+
+	trace2_region_enter("maintenance", "migrate_pack", the_repository);
+
+	basename = xstrndup(pack_filename, strlen(pack_filename) - 5 /*.pack*/);
```

Comment:
An alternative that would (at least to me) look more consistent with Git's opinionated source code style would be to set `size_t len = strlen(pack_filename) - 5; /* .pack */` and then using `strbuf_addf(&src, "%s/%.*s.idx", srcdir, len, pack_filename);` below, i.e. avoid copying the basename explicitly.

An even more consistent way would be to populate `src` and `dst` with the full path to the pack file, then use `strbuf_strip_suffix()` on it and then store the lengths and reset them via `strbuf_setlen()` instead of `strbuf_reset()`. Not sure whether you want to do that though?

Reply:
`"%s/%.*s.idx"` cannot be used when `len` is a `size_t`, unless we cast to `int`?

Reply:
Right. Better use the `strbuf_strip_suffix()` route, then?

Reply:
I'll just move to `strbuf_add`

Reply:
If you like the end result better, it is certainly fine by me.